src/goLanguageServer.ts

```diff
@@ -6,91 +6,118 @@
+	ctx.subscriptions.push(vscode.commands.registerCommand('go.languageserver.restart', () => {
+		return restartLanguageServer(ctx, parseLanguageServerConfig());
+	}));
 
-	// If the user has not enabled the language server,
-	// register the default language features and return.
-	if (!config.enabled) {
-		registerUsualProviders(ctx);
-		return;
+	const config = parseLanguageServerConfig();
+	if (!config || !config.enabled) {
+		return false;
 	}
+	const tool = getTool(config.serverName);
+	if (!tool) {
+		return false;
+	}
+	const update = await shouldUpdateLanguageServer(tool, config.path, config.checkForUpdates);
+	if (update) {
+		promptForUpdatingTool(tool.name);
+	}
+	// This function handles the case when the server isn't started yet,
+	// so we can call it to start the language server.
+	return restartLanguageServer(ctx, config);
+}
 
-	// The user has opted into the language server.
-	const languageServerToolPath = getLanguageServerToolPath();
-	const toolName = getToolFromToolPath(languageServerToolPath);
-	if (!toolName) {
-		// language server binary is not installed yet.
-		// Return immediately. The information messages such as
-		// offering to install missing tools, and suggesting to
-		// reload the window after installing the language server
-		// should be presented by now.
-		return;
+async function restartLanguageServer(ctx: vscode.ExtensionContext, config: LanguageServerConfig): Promise<boolean> {
+	// If the client has already been started, make sure to clear existing
+	// diagnostics and stop it.
+	if (languageClient) {
+		if (languageClient.diagnostics) {
+			languageClient.diagnostics.clear();
+		}
+		await languageClient.stop();
+		if (languageServerDisposable) {
+			languageServerDisposable.dispose();
+		}
 	}
-	const env = getToolsEnvVars();
 
-	// If installed, check. The user may not have the most up-to-date version of the language server.
-	const tool = getTool(toolName);
-	const update = await shouldUpdateLanguageServer(tool, languageServerToolPath, config.checkForUpdates);
-	if (update) {
-		promptForUpdatingTool(toolName);
+	// Check if we should recreate the language client. This may be necessary
+	// if the user has changed settings in their config.
+	if (!deepEqual(latestConfig, config)) {
+		// Track the latest config used to start the language server.
+		latestConfig = config;
+
+		// If the user has not enabled or installed the language server, return.
+		if (!config.enabled || !config.path) {
+			return false;
+		}
+		rebuildLanguageClient(config);
 	}
 
-	const c = new LanguageClient(
-		toolName,
+	languageServerDisposable = languageClient.start();
+	ctx.subscriptions.push(languageServerDisposable);
+
+	return true;
+}
+
+function rebuildLanguageClient(config: LanguageServerConfig) {
+	// Reuse the same output channel for each instance of the server.
+	if (!serverOutputChannel) {
+		serverOutputChannel = vscode.window.createOutputChannel(config.serverName);
```

Comment:
This looks new, are we creating a new output channel for gopls?
Doesnt gopls get a dedicated channel already?

Reply:
It gets created automatically so there is one already, but I figured we should reuse it across different `gopls` instances if it gets restarted.

Reply:
Can we not share the one that VS Code creates for it? My concern is that we will now have 2 output channels for gopls. One being created here, the other that we already have

Reply:
The VS Code language client creates one automatically if you don't set it, but by setting it in the language client object it won't, so we're fine.

Reply:
Ah! got it